xarray/convert.py

```diff
@@ -49,3 +64,130 @@ def set_cdms2_attrs(var, attrs):
+        coord_args = _get_iris_args(coord.attrs)
+        coord_args['var_name'] = coord_name
+        axis = None
+        if coord.dims:
+            axis = dataarray.get_axis_num(coord.dims)
+        if coord_name in dataarray.dims:
+            iris_coord = iris.coords.DimCoord(coord.values, **coord_args)
+            dim_coords.append((iris_coord, axis))
+        else:
+            iris_coord = iris.coords.AuxCoord(coord.values, **coord_args)
+            aux_coords.append((iris_coord, axis))
+
+    args = _get_iris_args(dataarray.attrs)
+    args['var_name'] = dataarray.name
+    args['dim_coords_and_dims'] = dim_coords
+    args['aux_coords_and_dims'] = aux_coords
+    if 'cell_methods' in dataarray.attrs:
+        args['cell_methods'] = parse_cell_methods(
+            dataarray.name, dataarray.attrs['cell_methods'])
+
+    cube = iris.cube.Cube(dataarray.to_masked_array(), **args)
+    return cube
+
+
+def _iris_obj_to_attrs(obj):
+    """ Return a dictionary of attrs when given a Iris object
+    """
+    attrs = {'standard_name': obj.standard_name,
+             'long_name': obj.long_name}
+    if obj.units.calendar:
+        attrs['calendar'] = obj.units.calendar
+    if obj.units.origin != '1':
+        attrs['units'] = obj.units.origin
+    attrs.update(obj.attributes)
+    return dict((k, v) for k, v in attrs.items() if v is not None)
+
+
+def _iris_cell_methods_to_str(cell_methods_obj):
+    """ Converts a Iris cell methods into a string
+    """
+    cell_methods = []
+    for cell_method in cell_methods_obj:
+        names = ''.join(['{}: '.format(n) for n in cell_method.coord_names])
+        intervals = ' '.join(['interval: {}'.format(interval)
+                              for interval in cell_method.intervals])
+        comments = ' '.join(['comment: {}'.format(comment)
+                             for comment in cell_method.comments])
+        extra = ' '.join([intervals, comments]).strip()
+        if extra:
+            extra = ' ({})'.format(extra)
+        cell_methods.append(names + cell_method.method + extra)
+    return ' '.join(cell_methods)
+
+
+def from_iris(cube):
+    """ Convert a Iris cube into an DataArray
+    """
+    import iris.exceptions
+    name = cube.var_name
+    dims = []
+    for i in xrange(cube.ndim):
+        try:
+            dim_coord = cube.coord(dim_coords=True, dimensions=(i,))
+            dims.append(dim_coord.var_name)
+        except iris.exceptions.CoordinateNotFoundError:
+            dims.append("dim_{}".format(i))
+
+    # dims = [dim.var_name for dim in cube.dim_coords]
+    # if not dims:
+    #     dims = ["dim{}".format(i) for i in range(cube.data.ndim)]
+    coords = OrderedDict()
+
+    for coord in cube.coords():
+        coord_attrs = _iris_obj_to_attrs(coord)
+        coord_dims = [dims[i] for i in cube.coord_dims(coord)]
+        if coord_dims:
+            coords[coord.var_name] = (coord_dims, coord.points, coord_attrs)
```

Comment:
We usually just raise ValueError for cases like this where we cannot handle
the input.

On Wed, Aug 10, 2016 at 2:01 PM, Neil Parley <notifications@github.com>
wrote:

> In xarray/convert.py
> <https://github.com/pydata/xarray/pull/814#discussion_r74329887>:
>
> > +        try:
> > +            dim_coord = cube.coord(dim_coords=True, dimensions=(i,))
> > +            dims.append(dim_coord.var_name)
> > +        except iris.exceptions.CoordinateNotFoundError:
> > +            dims.append("dim_{}".format(i))
> > +
> > +    # dims = [dim.var_name for dim in cube.dim_coords]
> > +    # if not dims:
> > +    #     dims = ["dim{}".format(i) for i in range(cube.data.ndim)]
> > +    coords = OrderedDict()
> > +
> > +    for coord in cube.coords():
> > +        coord_attrs = _iris_obj_to_attrs(coord)
> > +        coord_dims = [dims[i] for i in cube.coord_dims(coord)]
> > +        if coord_dims:
> > +            coords[coord.var_name] = (coord_dims, coord.points, coord_attrs)
>
> Is there an xarray exception that might make sense for the error? Or just
> nameerror?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pydata/xarray/pull/814/files/338ef6b7242c06d5107d2224c1df4cdb2a4b951c#r74329887>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKS1vRRJWIawQDcDJYZxXjQEcbQJIjkks5qejw5gaJpZM4H9jGH>
> .
>